Record polling refactor

mwbenowitz opened the pull request:
This update removes the Express application that wrapped the accessibility code generation and replaces it with an app that uses long-polling to monitor an SQS feed for report generation requests. When received the app generates the report (using the ACE accessibility reporting tool) and then places the result in a Kinesis stream, the same as before.

Several further changes had to be changed to accommodate the restrictions of the SQS queue. Most significantly, the buffer of the ePub archives cannot be directly passed to the application. Instead a S3 key reference is passed in the SQS messages and the application utilizes this to create a local copy of the ePub file that is then fed to the accessibility report generation method.

As part of this update, the test suite has been improved and expanded and every method fully documented in a jsDoc compliant format.

mwbenowitz committed:
SFR-367 Convert app to poller instead of API

The previous iteration of this tool was configured as an API to
respond to request to generate accessibility reports. This proved
to be very difficult to scale, because even when it was run in
an asynchronous mode it would still fail to acknowledge requests
and cause a cascade of errors in the ingest pipeline.

This version instead utilizes an SQS queue to process records,
allowing it to be entirely decoupled from the function(s) that
invoke it. This involved swapping the `Express` application for
an application based on the `sqs-consumer` library, which polls
a provided queue for new records.

The internal logic of this service remains largely the same, however
some additional logic is necessary because almost no ePub files will
fit in the 256 kB limit of SQS messages. Instead a S3 key is passed
in the message for the ePub to be rated. This file is then pulled
from the ePub bucket and temporarily stored while it is scored.

This commit includes all of the new logic for process the reports and
placing them in a Kinesis stream for processing. Another commit
will include updates to the internal documentation, README and test
suite.

mwbenowitz committed:
SFR-367 Add docstrings to methods

This adds docstrings and comments to the app

mwbenowitz committed:
SFR-367 Linting syntax fixes

mwbenowitz committed:
SFR-367 Unit testing and associated improvements

This commit extends and updates the test suite to provide coverage
for all components of the application with the exception of the app
wrapper itself.

Several minor adjustments were made to some functions to allow for
easier testing and in the case of the env variable setting function,
updated to force the updated of variables that already exist in
`process.env`

mwbenowitz committed:
SFR-367 Testing lint fixes

mwbenowitz committed:
SFR-367 Update documentation

mwbenowitz committed:
SFR-367 Update ePub download for better error handling

Potential error cases were being missed in the `downloadEpubFile`
method as error catching was only implemented on the `pipe` call.
This atomizes this process more and allows for better granularity
in catching and handling errors with files coming from S3

gkallenberg commented on README.md:
```diff
@@ -1,30 +1,36 @@
-- NGINX is used as a reverse proxy. This should be installed and configured to proxy requests to port 3000
+
 - pm2 is used to daemonize the API. This should be installed and use to start/run the node app.
 - Chromium is a dependency of the Ace Accessibility Report. npm will install this package but depending on the linux distro several dependencies may be missing. If that is the case you may need to install the libX libraries or others. It is difficult to predict. This can be partially solved by installing chromium via yum/apt-get which will pull in at least some of those dependencies if not all. * ...
```
...package, but, depending on the Linux distro, ...
If that is the case, ...
...Chromium...
Change `dependcies` to `dependencies.` Add a period at the end.

gkallenberg commented on README.md:
````diff
@@ -1,30 +1,36 @@
+```
 
 ## Environment
 This API can be deployed in an ec2 instance, and depending on the AMI used several dependencies may need to be installed/configured
````
...EC2 instance and, depending on the AMI used, ...

gkallenberg commented on docs/rationale.md:
```diff
@@ -1,30 +1,43 @@
 
 ## Daisy Ace
+
 The [EPUB Accessibility report generator](https://daisy.github.io/ace/) was designed as a command-line tool only, without documented support for use as a npm package in another project (though it is written in node.js). In order to use it for this service (on demand in response to a Kinesis stream) it was necessary to use it as a package. The Report generator works by splitting a zipped EPUB file ...
```
Capitalize Chromium.

gkallenberg commented on docs/rationale.md:
```diff
@@ -1,30 +1,43 @@
 
 ## AWS Lambda
+
 The main components of the ResearchNow data ingest pipeline are running in Lambda instances to allow for an easily scaleable and maintainable pipeline. The function that takes EPUB files and stores them in s3 takes the zipped EPUB files, explodes them into their components and stores them as both zipped and exploded versions. This runs smoothly but encounters several problems.
```
Capitlize S3.

gkallenberg commented on docs/rationale.md:
```diff
@@ -1,30 +1,43 @@
+
 The main components of the ResearchNow data ingest pipeline are running in Lambda instances to allow for an easily scaleable and maintainable pipeline. The function that takes EPUB files and stores them in s3 takes the zipped EPUB files, explodes them into their components and stores them as both zipped and exploded versions. This runs smoothly but encounters several problems.
 
 First, the package size of Lambdas is generally limited to 50MB for the most common upload methods. Including the Ace Report tool in this Lambda increased the size of the package to 200MB+ due to various dependencies. This could be mitigated by uploading the package via s3, however that is not ideal from a maintainability standpoint.
```
Capitalize S3.

gkallenberg commented on docs/rationale.md:
```diff
@@ -1,30 +1,43 @@
 
 First, the package size of Lambdas is generally limited to 50MB for the most common upload methods. Including the Ace Report tool in this Lambda increased the size of the package to 200MB+ due to various dependencies. This could be mitigated by uploading the package via s3, however that is not ideal from a maintainability standpoint.
 
 Second, the startup time of the Lambda is severly affected by inlcuding the Ace Reporting tool. It adds several seconds to the startup of each Lambda instance (largely due to Chromium) and its use is therefore generally not encouraged. One of the main reasons an ec2 instance was created was because it is always available and eliminates the need for startup on each call.
```
Capitalize EC2.

gkallenberg commented on docs/rationale.md:
```diff
@@ -1,30 +1,43 @@
 
 ## Chromium
+
 As a dependency of the Ace Reporter Chromium adds over 80MB in requirements. Additionally, depending on the flavor of linux it is being deployed on it may be missing several key libraries (mainly libX and its dependencies), which complicates deployment in many enviroments. There are several projects to create node and/or Lambda centric distributions of Chromium, which may provide a path forward f ...
```
Capitalize Linux and add a comma after 'being deployed on'

gkallenberg commented on docs/rationale.md:
```diff
@@ -1,30 +1,43 @@
 
 ## EC2 Size/Responsiveness
+
 As it stands now this service is running on a t3.small ec2 instance. This was selected as the t3 isntances offer more vCPU cores at lower levels than the standard t2 instances. This allows for better parallel performance as multiple EPUBs can be processed at once (ideally having a 3+ core machine would process a full batch in parallel, but having 2 cores seems to work well at the moment). The Ace ...
```
Capitalize EC2.
Change 'isntances' to 'instances'

gkallenberg commented on docs/rationale.md:
```diff
@@ -1,30 +1,43 @@
 
 ## Future Work
+
 There are several potential solutions to these questions. The most comprehensive would be to either eliminate the Chromium dependency of the Ace Reporting tool, however that seems unlikely as it is a key part of their reporting tool. Replacing it with a Lambda compatible Chromium package is perhaps more feasible and will be on the To-Do list for future tasks.
```
Remove 'either' here since your spelling out another option below.
Use '...tool; however, ...' instead?

gkallenberg commented on docs/rationale.md:
```diff
@@ -1,30 +1,43 @@
+
 There are several potential solutions to these questions. The most comprehensive would be to either eliminate the Chromium dependency of the Ace Reporting tool, however that seems unlikely as it is a key part of their reporting tool. Replacing it with a Lambda compatible Chromium package is perhaps more feasible and will be on the To-Do list for future tasks.
 
 Fine tuning the ec2 instance for performance/cost is another option as having a standalone API offers several advantages, such as being able to initiate/generate Accessibility Reports from arbitrary points in the ResearchNow pipeline. For example, an automated tool that updates/fixes epubs could generate new reports without re-implementing any code around them.
```
Capitalize EC2.

gkallenberg commented on index.js:
```diff
@@ -1,14 +1,52 @@
+import { Consumer } from 'sqs-consumer'
+import { runAccessibilityReport } from './src/accessibility_report.js'
+import { resultHandler } from './src/kinesis_out.js'
+import { setEnv } from './src/env_config.js'
```
Are you setting or getting?

mwbenowitz replied:
hmm, good question. It's getting the vars from a `.env` file for the local environment and setting them. Maybe `loadEnv` is a more accurate name?

gkallenberg replied:
That sounds like a better verb to use.

gkallenberg commented on src/accessibility_report.js:
```diff
@@ -1,102 +1,178 @@
+/*
+ * @async
+ * Accesses and stores an .epub file (zipped archive) from the S3 bucket defined
+ * in the current environement variables.
```
environment (sp)

mwbenowitz committed:
SFR-367 Fix self-documenting comments

The comments placed in the source files was not being recognized
by the `jsdoc` documentation generator. This resolves the issue

gkallenberg commented on src/accessibility_report.js:
```diff
@@ -1,102 +1,182 @@
 }
 
+/**
+ * Generates a summary score based of the number and severity of the violations
```
based on?

gkallenberg commented on test/accessibilityReport.test.js:
```diff
@@ -150,12 +189,12 @@ describe('Accessibility Checker [accessibilityCheck.js]', () => {
+        moderate: 69,
+        minor: 0
+      }
       let score = AccessibilityChecker.calculateScore(violations)
```
`violations` and `score` seem to be `const` values, no?

gkallenberg reviewed (commented):
Looking good. Just a few comments.

mwbenowitz committed:
SFR-367 Fix spelling errors and method name

This corrects several spelling errors and other typos. It also
changes the name of the environment variable handler from `setEnv`
to `loadEnv` to hopefully make the purpose/function of the
method more clear.

mwbenowitz commented:
Made the corrections found here (after installing a spellchecker 🙂) and renamed the function `setEnv` to `loadEnv`
